################################################
## Three-Dimensional Human Alveolar Stem Cell Culture Models Reveal Infection Response to SARS-CoV-2
## Cell Stem Cell, October 21, 2020, DOI:https://doi.org/10.1016/j.stem.2020.10.004
## 
## Scripts for Seurat&SingleCellExperiment Objects generation of single cell RNA sequencing
## If you want to make figures, please see 03_Figure_scRNAseq.R in https://github.com/ju-lab/SARS-CoV-2_alveolar_organoids
##
## Scripts have been posted in https://github.com/ju-lab/SARS-CoV-2_alveolar_organoids
## Feature_bc_matrix data and Seurat/SingleCellExperiment objects have been uploaded in https://www.synapse.org/#!Synapse:syn22146555/files/
## Bam files for the five single cell data have been deposited in the European Genome-Phenome Archive (EGA) with accession ID EGAS00001004508 
## 3rd Dec. 2020. by Su Yeon Kim, Jeonghwan Youk
################################################

##-------------------------------------------
## Directory setting
##-------------------------------------------
#working_dir <- "/home/users/jhyouk/13_SARS_COV_2/github_script"
working_dir <- "/PATH_To_workding_directory"
setwd(working_dir)

dat_dir <- file.path(working_dir,"R_data/")
if(!dir.exists(dat_dir)) dir.create(dat_dir)


##-------------------------------------------
## Load libraries and source files
##-------------------------------------------
require(readxl)
require(RColorBrewer)

require(Biobase)
require(GEOquery)

library(foreach)
library(doParallel)

require(SingleCellExperiment)
require(scran)
require(scater)
require(grid)
require(ggplot2)
require(gridExtra)
require(pheatmap)

require(Seurat)

require(DESeq2)
require(apeglm)

require(SC3)
require(TSCAN)
require(M3Drop)
require(tidyverse)


##-------------------------------------------
## shared color settings
##-------------------------------------------

cols.cond <- c("olivedrab2", "goldenrod2", "orangered1", "red3", "violetred3")
cols.clusterV3 <- c(brewer.pal(12, "Paired")[c(2, 4)],
                    "darkorange1",
                    brewer.pal(12, "Paired")[c(12)],
                    brewer.pal(12, "Paired")[c(10)],
                    "magenta", "navyblue")
cols.sc3_clusters <- c(brewer.pal(8, "Accent")[1:7], brewer.pal(12, "Paired")[c(10, 6)])
cols.levelSARS <- c(gray(0.8), brewer.pal(9, "OrRd")[c(5, 7, 9)])

##-------------------------------------------
## genes of interest 
##-------------------------------------------

lgenes <- list(GenesA=c("SFTPB","SFTPC","NKX2-1",  ## AT2 marker
                        "IFI44L","IFI6","MX1",  ## Interferon
                        "GADD45B","TNFAIP3", # APOPTOSIS
                        "ACE2","TMPRSS2"),  # SARS-CoV-2 related
               GenesB = c("PDPN","AGER","CLIC5"), # AT1 marker
               GenesC = c("SOX2","KRT17"),  ## sick
               GenesD = c("KRT5", "TP63"),  ## basal
               GenesE = c("MSR1","MKI67", "KIT"))  ## others


##-------------------------------------------
## key factors for cells 
##-------------------------------------------

lsinfos <- list(SARSCoV2= c("SARS-CoV-2", "Frac-SARS-CoV-2", "Level-SARS-CoV-2"),
                UMIcounts = c("Total-UMI", "log2-Total-UMI", "Human-UMI", "vcounts.sarsgene"),
                Clustering = c("Condition", "ClusterrefitV3Fac", "sc3_9_clusters"))








##-------------------------------------------
## Load initial filtered_feature_bc_matrix files which were generated by cellranger.
## You can download raw "filtered_feature_bc_matrix" of five groups in https://www.synapse.org/#!Synapse:syn22146555/files/
##-------------------------------------------

ldirs <- c(s00h_0.0moi="/home/users/jhyouk/13_SARS_COV_2/scRNAseq/sc_control/outs/filtered_feature_bc_matrix",
           s04h_0.1moi="/home/users/jhyouk/13_SARS_COV_2/scRNAseq/MOIpoint1/sars_4hpi/outs/filtered_feature_bc_matrix",
           s18h_0.1moi="/home/users/jhyouk/13_SARS_COV_2/scRNAseq/MOIpoint1/sars_18hpi/outs/filtered_feature_bc_matrix",
           s60h_0.1moi="/home/users/jhyouk/13_SARS_COV_2/scRNAseq/MOIpoint1/sars_60hpi/outs/filtered_feature_bc_matrix",
           s60h_1.0moi="/home/users/jhyouk/13_SARS_COV_2/scRNAseq/sc_sars/outs/filtered_feature_bc_matrix")

lscdats <- lapply(ldirs, function (mydir) Read10X(data.dir = mydir))
lcounts <- lapply(lscdats, function (sc) as.matrix(sc)) ## convert as usual matrix 

## Merge the five matrices
for(k in 1:length(lcounts)) {
  colnames(lcounts[[k]]) <- paste0(names(lcounts)[k], "__", colnames(lcounts[[k]]))
}
counts.comb <- do.call(cbind, lcounts)

##-------------------------------------------
## Create column information
##-------------------------------------------

sinfo <- data.frame(Cell=colnames(counts.comb),
                    SampleID=gsub("(.*)__(.*)", "\\1", colnames(counts.comb)), 
                    stringsAsFactors = F)
sinfo$Condition <- factor(gsub("s", "", sinfo$SampleID),
                          levels=c("00h_0.0moi",
                                   "04h_0.1moi", 
                                   "18h_0.1moi",
                                   "60h_0.1moi", 
                                   "60h_1.0moi"),
                          labels=c("0h_0.0moi",
                                   "4h_0.1moi", 
                                   "18h_0.1moi",
                                   "60h_0.1moi", 
                                   "60h_1.0moi"))

vcounts.tot <- colSums(counts.comb)
vcounts.sarsgene <- counts.comb["SARS-CoV-2",]
vcounts.hmgenes <- vcounts.tot - vcounts.sarsgene
vcounts.MT <- colSums(counts.comb[grep("^MT-", rownames(counts.comb)),])

frac.sarsgene <- vcounts.sarsgene/vcounts.tot
frac.MT.overhmgenes <- vcounts.MT / vcounts.hmgenes

sinfo$vcounts.tot <- vcounts.tot
sinfo$vcounts.sarsgene <- vcounts.sarsgene
sinfo$vcounts.hmgenes <- vcounts.hmgenes
sinfo$vcounts.MT <- vcounts.MT
sinfo$frac.sarsgene <- frac.sarsgene
sinfo$frac.MT.overhmgenes <- frac.MT.overhmgenes

##-------------------------------------------
## Add row data (gene information)
##-------------------------------------------

fdat <- data.frame(Gene=rownames(counts.comb), stringsAsFactors = F)
fdat$IS_MT <- grepl("^MT-", fdat$Gene)


##-------------------------------------------
## Create singlecell experiment object 
##-------------------------------------------

sce.allgenes.raw <- SingleCellExperiment(assays = list(counts=counts.comb),
                                         colData=sinfo, 
                                         rowData=fdat)


##-------------------------------------------
## remove SARS-gene to leave human genes only for normalization
##-------------------------------------------

sce.hmgenes.raw <- sce.allgenes.raw[rowData(sce.allgenes.raw)$Gene!="SARS-CoV-2",]

totscale <- 10000
mylognorm <- log2 (t(t(counts(sce.hmgenes.raw))/colSums(counts(sce.hmgenes.raw))*totscale)+ 1)
assays(sce.hmgenes.raw)[["logcounts"]] <- mylognorm

## normalize sarsgene in the same way 
sce.hmgenes.raw$lognorm.sargene <- log2(sce.hmgenes.raw$vcounts.sarsgene / colSums(counts(sce.hmgenes.raw))*totscale+1)


##-------------------------------------------
## Remove cells with fraction of mitochondrial genes >= 0.25
##-------------------------------------------

sce.hmgenes.blMT025 <- sce.hmgenes.raw[,sce.hmgenes.raw$frac.MT.overhmgenes < 0.25]

  

##-------------------------------------------
## Convert SingleCellExperiment object to Seurat object  & initial clustering
##-------------------------------------------

sro <- as.Seurat(sce.hmgenes.blMT025, 
                 counts = "counts",
                 data = "logcounts",
                 assay = "RNA",
                 project = "SingleCellExperiment")

sro <- FindVariableFeatures(sro, selection.method = "vst", nfeatures = 3000)
sro@assays$RNA@scale.data <- sro@assays$RNA@data


sro <- RunPCA(sro, 
              features = VariableFeatures(sro)[1:3000], 
              npcs = 50, nfeatures.print = 30)
sro <- RunUMAP(sro, dims=1:20)

DimPlot(sro, reduction = "pca", 
        group.by = "Condition", cols=cols.cond  )
DimPlot(sro, reduction = "umap", 
        group.by = "Condition", cols=cols.cond )


## Clustering
sro <- FindNeighbors(sro, dims = 1:20)
sro <- FindClusters(sro, resolution = 0.2)

sroF <- sro
sroF@meta.data[,"SARS-CoV-2"]  <- sroF$lognorm.sargene
sroF@meta.data[,"Total-UMI"] <- sroF$vcounts.tot
sroF@meta.data[,"log2-Total-UMI"] <- log2(sroF@meta.data$`Total-UMI`)


sroF$clusterID <-  Idents(sroF)
sroF$ClusterFac <- factor(sroF$clusterID, 
                          levels=c("3",
                                   "7",
                                   "8",
                                   "6",
                                   "5",
                                   "4",
                                   "1",
                                   "2",
                                   "0",
                                   "9")
                          ,
                          labels=c("cl3_0h_m0.0",
                                   "cl7_4h_m0.1_highSF",
                                   "cl8_18h_m0.1_highSF",
                                   "cl6_60h_m0.1_highSF",
                                   "cl5_60h_m1.0_lowCoV2",
                                   "cl4_60h_m1.0_highCoV2",
                                   "cl1_4h_m0.1_lowSF",
                                   "cl2_18h_m0.1_lowSF",
                                   "cl0_60h_m0.1_lowSF",
                                   "cl9_airways"))

vecgenes <- unique(unlist(lgenes)); vecgenes
FeaturePlot(sroF, vecgenes, ncol=5)

##-------------------------------------------
## Remove pathologic cell population (clusters with higher expression level of KRT17 & lower expression level of surfactant)
##-------------------------------------------

sroE <- sroF[, sroF$clusterID %in% c(3:9)] # remove cluster 0, 1, 2


##-------------------------------------------
## Re-do#1 variable feature selection & clustering 
##-------------------------------------------

sroE <- FindVariableFeatures(sroE, selection.method = "vst", nfeatures = 3000)

sroE@reductions <- list()  ## reset reductions
sroE <- RunPCA(sroE, 
                features = VariableFeatures(sroE), 
                npcs = 50, nfeatures.print = 30)
sroE <- RunUMAP(sroE, dims=1:20)

#FeaturePlot(sroE, reduction="umap", features = vecgenes, ncol=5)
sroE <- FindNeighbors(sroE, dims = 1:20)
sro.clust.0.1 <- FindClusters(sroE, resolution = 0.1)
DimPlot(sro.clust.0.1, reduction = "umap")


##-------------------------------------------
## Remove contaminated population (blood-origin cells) & re-do#2 variable feature selection & clustering
##-------------------------------------------

sro.clust.0.1$refit_clusterID <- Idents(sro.clust.0.1)
sroR <- sro.clust.0.1[, sro.clust.0.1$refit_clusterID != 7] # remove cluster' 7

sroR <- FindVariableFeatures(sroR, selection.method = "vst", nfeatures = 3000)
sroR@reductions <- list()  ## reset reductions

sroR <- RunPCA(sroR, 
                features = VariableFeatures(sroR), 
                npcs = 50, nfeatures.print = 30)

sroR <- RunUMAP(sroR, dims=1:30)
DimPlot(sroR, reduction="umap", group.by = "Condition", cols=cols.cond )


sroR <- FindNeighbors(sroR, dims = 1:20)
sroR.refitV3 <- FindClusters(sroR, resolution = 0.1)
DimPlot(sroR.refitV3, reduction = "umap")

sroR.refitV3@meta.data[,"Human-UMI"] <- sroR.refitV3$vcounts.hmgenes
sroR.refitV3@meta.data[,"Frac-SARS-CoV-2"] <- sroR.refitV3$frac.sarsgene



##----------------------------------------------------------
## check unsupervised clustering results
##----------------------------------------------------------
sroR.refitV3$refitV3_clusterID <- Idents(sroR.refitV3)
sroR.refitV3$ClusterrefitV3Fac <- factor(sroR.refitV3$refitV3_clusterID,
                                         levels=c("0", "4", "5", "2", "1", "3", "6"),
                                         labels=c("rcl0_0h_m0.0",
                                                  "rcl4_4h_m0.1",
                                                  "rcl5_18h_m0.1",
                                                  "rcl2_60h_m0.1",
                                                  "rcl1_60h_m1.0_lowCoV2",
                                                  "rcl3_60h_m1.0_highCoV2.mix",
                                                  "rcl6_airway-like"))



##========================================================
## Convert Seurat object to SingleCellExperiment object
## Then, clustering using SC3 package
##========================================================

sroT <- sroR.refitV3
sroT <- FindVariableFeatures(sroT, selection.method = "vst", nfeatures = 10000)
sceT <- as.SingleCellExperiment(sroT)
sceT@metadata <- list(VariableFeatures=VariableFeatures(sroT))

counts(sceT) <- as.matrix(counts(sceT)) # make count as usual matrix 
rowData(sceT)$feature_symbol <- rownames(sceT) ## feature_symbol required for SC3 clustering



##--------------------------------------------------------
## Perform clustering with ks = 8 to 9
## In case of uploaded RDS files, clustering was performed with ks = 5 to 11
##--------------------------------------------------------
sceT.g5k <- sceT[ sceT@metadata$VariableFeatures[1:5000], ] # Use most variable 5000 features
sceT.g5k <- sc3(sceT.g5k, ks=8:9, biology=TRUE, n_cores=5)

sroP<-sroR.refitV3
sceP<-sceT.g5k

##-----------------------------------------------------------
## Categorize expression level of SARS-CoV-2
## break points: 1, 3, 9, and 18
##-----------------------------------------------------------

sroP@meta.data[,"Level-SARS-CoV-2"] <- cut(sroP@meta.data[,"SARS-CoV-2"], 
                                           breaks=c(0, 1, 3, 9, 18), 
                                           include.lowest = T, right=FALSE)

sroP@meta.data[,"IntLevel-SARS-CoV-2"] <- as.integer(sroP@meta.data[,"Level-SARS-CoV-2"])



##-----------------------------------------------------------
## Add SC3 clustering information to Seurat object and remove temporary fields in Seurat object
##-----------------------------------------------------------

sroP@meta.data$sc3_9_clusters <- sceP$sc3_9_clusters
sroP@meta.data$sc3_9_log2_outlier_score <- sceP$sc3_9_log2_outlier_score
sroP@assays$RNA@meta.features$IsTopVarFeat5k <- ifelse(rownames(sroP) %in% rownames(sceP), 
                                                       "In_TopVarFeat5k",
                                                       "Not_TopVarFeat5k")

## remove temporary fields
lf.toexcl <- colnames(sroP@meta.data)[grep ("comb.", colnames(sroP@meta.data))]
sroP@meta.data <- sroP@meta.data[,!(colnames(sroP@meta.data) %in% lf.toexcl)]
lf.copy <- c("sc3_gene_filter",
             "sc3_9_markers_clusts",
             "sc3_9_markers_padj",
             "sc3_9_markers_auroc",
             "sc3_9_de_padj")
for(f in lf.copy){ 
  sroP@assays$RNA@meta.features[,f] <- rowData(sceP)[match(rownames(sroP), rownames(sceP)), f]
}

lf.toexcl2 <- colnames(colData(sceP))[grep ("comb.|sc3_[5678]|sc3_1[01]", colnames(colData(sceP)))]
colData(sceP) <- colData(sceP)[, !(colnames(colData(sceP)) %in% lf.toexcl2)]
lf.copy2 <- c("Human-UMI",
              "Frac-SARS-CoV-2",
              "Level-SARS-CoV-2",
              "IntLevel-SARS-CoV-2")
for(f in lf.copy2) colData(sceP)[,f] <- sroP@meta.data[,f]

lf.toexcl3 <- colnames(rowData(sceP))[grep ("comb.|sc3_[5678]|sc3_1[01]", colnames(rowData(sceP)))]
rowData(sceP) <- rowData(sceP)[, !(colnames(rowData(sceP)) %in% lf.toexcl3)]


##==========================================
## save objects as final 
##==========================================
sroF <- sroP
sceF <- sceP

saveRDS(sroF, file.path(dat_dir, 
                          "SeuratObj_RefitV3.rmlowSF.rmBlood_UMAPpc30_clust0.1_FINAL_cleaned_new.RDS"))
saveRDS(sceF, file.path(dat_dir, 
                                "SCE_rmlowSF.rmBlood_g5k_SC3_k9_trained_FINAL_new.RDS"))
  
                












##++++++++++++++++++++++++++++++++++++++++++
## !!!!Optional fields!!!!
## Predict clusters in the remaing cells which were not randomly selected in SC3 clustering above.
##++++++++++++++++++++++++++++++++++++++++++

sce <- sceP
sce <- sc3_run_svm(sce, ks = 9)
metadata(sce)$sc3$svm_train_inds <- NULL
sce <- sc3_calc_biology(sce, ks = 9)

col_data <- colData(sce)
table(colData(sce)$ClusterrefitV3Fac, colData(sce)$sc3_9_clusters, useNA='ifany')